webpack/components/Table/TableWrapper.js

```diff
@@ -0,0 +1,108 @@
+import React, { useEffect, useState } from 'react';
+import {
+  Table,
+  TableHeader,
+  TableBody,
+} from '@patternfly/react-table';
+import {
+  Pagination,
+} from '@patternfly/react-core';
+import PropTypes from 'prop-types';
+import { useDispatch } from 'react-redux';
+import { STATUS } from 'foremanReact/constants';
+import { useForemanSettings } from 'foremanReact/Root/Context/ForemanContext';
+
+import EmptyStateMessage from './EmptyStateMessage';
+import Loading from './Loading';
+
+/* Patternfly 4 table wrapper */
+const TableWrapper = ({
+  status, cells, rows, error, emptyTitle, emptyBody, metadata, fetchItems, ...extraTableProps
+}) => {
+  const dispatch = useDispatch();
+  const { foremanPerPage = 20 } = useForemanSettings();
+  // setting pagination to local state so it doesn't disappear when page reloads
+  const [perPage, setPerPage] = useState(foremanPerPage);
+  const [page, setPage] = useState(1);
+  const [total, setTotal] = useState(0);
+
+  useEffect(() => {
+    const { total: newTotal, page: newPage, per_page: newPerPage } = metadata;
+    if (newTotal) setTotal(parseInt(newTotal, 10));
+    if (newPage) setPage(parseInt(newPage, 10));
+    if (newPerPage) setPerPage(parseInt(newPerPage, 10));
+  }, [metadata]);
+
+  const buildMainTable = () => {
```

Comment:
This is a function that returns JSX, therefore it's a React component and should be refactored as such (name should be TitleCase; render as regular JSX instead of inside brackets; etc.)  I've never liked the pattern of building the JSX to return within a React component by using helper functions that return JSX.  Instead let's use the amazing composability that JSX gives to us natively.  (No need to extract to a separate file, though.. I'm fine with multiple small components living in the same file if it makes sense.)

Reply:
I agree in principle, but I went with this way because I'm not crazy about the two alternatives:
- Put inside JSX - The logic for this many conditions is harder to read (IMO) when all in JSX and needs to be escaped
- Move to separate component - having to do more prop drilling by repeating the props and passing them once again

I'm not opposed to either of these as things grow, but I just don't see enough issue with the helper function to go with an alternative as things are right now. I'll definitely update the name to TitleCase as you suggested and maybe we can discuss the structure more as things progress?

Reply:
```js
{MainTable()}
```
can become 
```js
<MainTable />
```
and it just works.  Since the component is defined from within the same scope, it has access to `status`, `error`, `emptyTitle`, and `emptyBody` without explicitly passing them.  Yay JavaScript!

Reply:
Thanks, I updated. I figured you meant move it outside the parent function so was confused, but it makes more sense now :)